### ISSUE TITLE
ENH: allow the `set...` methods to be used as context managers

### ISSUE THREAD

**Issue by ianhi** [pymmcore-plus/pymmcore-plus#118]
A nice convenience would be if all of the various `set` methods could be used as context managers.

Something like:

```python
core.getAutoShutter() # True

with core.setAutoShutter(False):
   with core.setShutterOpen(True):
      ....
core.getAutoShutter() # True
```

or for multple states:

```python
with core.setContext('autoShutter':False, 'ShutterOpen':True):
    ...
```

**Comment by tlambert03** [comment]
took a quick and dirty crack at this with something like:
```py
for name in dir(CMMCorePlus):
    getter_name = f'get{name[3:].title()}'
    if name.startswith('set') and hasattr(CMMCorePlus, getter_name):
        print(name)
        setter = getattr(CMMCorePlus, name)
        getter = getattr(CMMCorePlus, getter_name)

        @contextmanager
        @wraps(setter)
        def context_setter(self, *args, **kwargs):
            before = getter()
            setter(self, *args, **kwargs)
            try:
                yield
            finally:
                setter(self, before)

        setattr(CMMCorePlus, name, context_setter)
```

but ran into swig limitations:
```
AttributeError: cannot modify read-only attribute 'CMMCorePlus.setExposure'
```

so it looks like we'd need to do this explicitly/exhaustively for every setter

**Comment by ianhi** [comment]
That functions looks like magic to me. How would it also preserve the non-context usage?

> but ran into swig limitations:

I wonder if we couldn't make some changes to the pymmcore swig definitions to make this possible.

**Comment by ianhi** [comment]
a convenience method could look like this:

```python
from contextlib import contextmanager
@contextmanager
def setContext(**kwargs):
    orig_values = {}
    try:
        for name,v in kwargs.items():
            name = name[0].upper() + name[1:]
            try:
                orig_values[name] = getattr(core, f"get{name}")()
                getattr(core, f"set{name}")(v)
            except AttributeError:
                pass
        yield
    finally:
        for k, v in orig_values.items():
            try:
                getattr(core, f"set{k}")(v)
            except AttributeError:
                pass
```

**Comment by tlambert03** [comment]
> How would it also preserve the non-context usage?

sorry, you're right, we can't use `contextlib.contextmanager`, since it doesn't separate init from `__enter__`.  i.e. with this:
```python
@contextmanager
def f():
    print(1)
    yield
    print(2)
```

calling `f()` won't print 1, but calling `with f(): ...` prints both 1 and 2

what we need here is something that prints 1 when called _without_ a with statement... and returns an object. The main point to recognize here is that if we want to overload the `setX` methods to let them be used as context managers, then you essentially _can't_ have them return `None` anymore; they have to return a context that can be entered with `__enter__`.

so, replace the `@contextmanager` bit in the above example with this

```python
class setSomething:
    def __init__(self, *args, **kwargs):
        self._before = getter_method()
        actual_setter_method(*args, **kwargs)
    def __enter__(self):
        return self
    def __exit__(self, *a):
        actual_setter_method(self._before)
```

and it should work


---

yes, the `setContext` method (which _isn't_ overloading a core function) is much easier and more wide open

**Comment by ianhi** [comment]
implemented in #120
